Source/HTTP/httpcall.cpp

```diff
@@ -137,6 +138,91 @@ HRESULT CALLBACK HC_CALL::PerfomAsyncProvider(XAsyncOp op, XAsyncProviderData co
     }
 }
 
+void CALLBACK HC_CALL::CompressRequestBody(void* c, bool canceled)
+{
+    PerformContext* context{ static_cast<PerformContext*>(c) };
+    HC_CALL* call{ context->call };
+
+    if (canceled)
+    {
+        XAsyncComplete(context->asyncBlock, E_ABORT, 0);
+        return;
+    }
+
+    if (call->traceCall)
+    {
+        HC_TRACE_INFORMATION(HTTPCLIENT, "HC_CALL::CompressRequestBody [ID %llu] Iteration %d", TO_ULL(call->id), call->m_iterationNumber);
+    }
+
+#if !HC_NOZLIB
+
+#if HC_PLATFORM == HC_PLATFORM_WIN32 || HC_PLATFORM == HC_PLATFORM_GDK
+    // Compress body before call if applicable
+    if (call->enableGzipCompression && call->compressionLevel != HCCompressionLevel::None)
+    {
+        HCHttpCallRequestBodyReadFunction clientRequestBodyReadCallback{ nullptr };
+        size_t requestBodySize{};
+        void* clientRequestBodyReadCallbackContext{ nullptr };
+        HRESULT hr = HCHttpCallRequestGetRequestBodyReadFunction(call, &clientRequestBodyReadCallback, &requestBodySize, &clientRequestBodyReadCallbackContext);
+
+        if (FAILED(hr) || !clientRequestBodyReadCallback)
+        {
+            HC_TRACE_ERROR(HTTPCLIENT, "CurlEasyRequest::ReadCallback: Unable to get client's RequestBodyRead callback");
+            return;
+        }
+        
+        size_t uncompressedRequestyBodySize = 0;
+        const size_t bufferSize = 128 * 1024;
+        uint8_t uncompressedRequestyBodyBuffer[bufferSize];
+
+        try
+        {
+            hr = clientRequestBodyReadCallback(call, 0, bufferSize, clientRequestBodyReadCallbackContext, uncompressedRequestyBodyBuffer, &uncompressedRequestyBodySize);
+
+            if (FAILED(hr))
+            {
+                HC_TRACE_ERROR_HR(HTTPCLIENT, hr, "CurlEasyRequest::ReadCallback: client RequestBodyRead callback failed");
+                return;
+            }
+        }
+        catch (...)
+        {
+            return;
+        }
+
+        http_internal_vector<uint8_t> compressedRequestBodyBuffer;
+
+        Compression::CompressToGzip(uncompressedRequestyBodyBuffer, static_cast<unsigned int>(uncompressedRequestyBodySize), call->compressionLevel, compressedRequestBodyBuffer);
+
+        auto requestCompressedBodyBytes = compressedRequestBodyBuffer.data();
+        auto requestCompressedBodySize = (uint32_t)compressedRequestBodyBuffer.size();
+
+        // Setting back to default read request body callback to be invoked by the Platform-specific code
+        call->requestBodyReadFunction = HC_CALL::ReadRequestBody;
+        call->requestBodyReadFunctionContext = nullptr;
+
+        // Directly setting compressed body bytes to HCCall
+        call->requestBodySize = requestCompressedBodySize;
+        call->requestBodyBytes.assign(requestCompressedBodyBytes, requestCompressedBodyBytes + requestCompressedBodySize);
```

Comment:
>call->requestBodyBytes.assign(requestCompressedBodyBytes, requestCompressedBodyBytes + requestCompressedBodySize); [](http://example.com/codeflow?start=8&length=114)

I believe this copies the request body buffer. You should be able to avoid that with std::move(compressedRequestBodyBuffer)

Reply:
Fixed